modules/linkPreview/logic/linkPreviewLogic.js

```diff
@@ -17,40 +17,59 @@ const { LinkPreview } = require('../../../models');
 const DomLoader = require('../utils/domLoader');
 const cache = require('../../cache/utils/cache');
 const imageLogic = require('../../media/logic/imageLogic');
+const { MESSAGE_QUEUES, MESSAGES } = require('../../queue/constants/queue');
+const queue = require('../../queue/logic/queueLogic');
+const CacheLogic = require('../../cache/logic/cacheLogic');
+
 const logger = require('../../../utils/logger')(module);
 
 lngDetector.setLanguageType('iso2');
 
-module.exports = class LinkPreviewLogic {
-  static async fetchAllLinkPreviews() {
+class LinkPreviewLogic {
+  constructor() {
+    queue.subscribeToMessage(MESSAGE_QUEUES.LINKPREVIEW, MESSAGES.LINKPREVIEW.COMMANDS.UPDATE_DB,
+      async message => {
+        await this.updateLinkpreviewDatabase();
+        await queue.deleteMessage(MESSAGE_QUEUES.LINKPREVIEW, message.id);
+      });
+  }
+
+  async fetchAllLinkPreviews() {
     return LinkPreview.findAll({ raw: true });
   }
 
-  static async fetchAllUrls() {
+  async fetchAllUrls() {
     return LinkPreview.aggregate('requestUrl', 'DISTINCT', { plain: false })
       .then(results => results.map(preview => preview.DISTINCT));
   }
 
-  // API Functions
-  static async getLinkPreview(req, res) {
-    const url = req.params.url ? req.params.url : req.query.url;
-    if (url) {
-      const result = await LinkPreview.findOne({ where: { requestUrl: url }, raw: true });
-      return result ? res.send(result) : res.sendStatus(404);
+  async updateLinkpreviewDatabase() {
+    const tips = await CacheLogic.getTips();
+    const previews = await this.fetchAllUrls();
+    const tipUrls = [...new Set(tips.filter(tip => tip.url).map(tip => tip.url))];
+
+    const difference = tipUrls.filter(url => !previews.includes(url));
+
+    await difference.asyncMap(async url => {
+      await this.generatePreview(url).catch(logger.error);
+    });
```

Comment:
You are right, this would be a lot quicker but it might have a couple of negative side-effects:
a) it consumes a lot more ram, maybe too much
b) if we query 10000 links from the same domain it could be considered a DDoS.
So the linearity protects us and everyone else from hostile actions for now. This could be changed in the future to allow X requests in parallel (maybe 10 or sth).